arch: deprecate `z_arch_esf_t` with `struct arch_esf`, introduce an arch-agnostic `exception.h` for it

ycsin opened the pull request:
RFC #73673

Fixes #73454

Unblocks #73587

- SOF
  - https://github.com/zephyrproject-rtos/sof/pull/46

[Migration guide preview](https://builds.zephyrproject.io/zephyr/pr/73593/docs/releases/migration-guide-3.7.html#kernel)

ycsin committed:
arch: define `struct arch_esf` and deprecate `z_arch_esf_t`

Make `struct arch_esf` compulsory for all architectures by
declaring it in the `arch_interface.h` header.

After this commit, the named struct `z_arch_esf_t` is only used
internally to generate offsets, and is slated to be removed
from the `arch_interface.h` header in the future.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
arch: Create `include/zephyr/arch/exception.h`

Create `zephyr/include/zephyr/arch/exception.h`, which will
redirect to the corrent architecture-specific exception header
based on Kconfig.

Some of the architectures define their esf struct in
architecture-specific `arch.h`, refactor them out into a
separate `exception.h`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
include: include `exception.h` in `fatal.h`

`fatal.h` has 2 functions that use the `z_arch_esf_t` type.

Include `exception.h`, which should have the `z_arch_esf_t`
defined.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

aescolar commented on include/zephyr/arch/exception.h:
```diff
@@ -0,0 +1,37 @@
+#elif defined(CONFIG_SPARC)
+#include <zephyr/arch/sparc/exception.h>
+#else
+#error "Unknown Architecture"
```
I wouldnt error out, as it would fail for out of tree architectures.
CC @daor-oti

ycsin replied:
Just want to point out that this isn't something new:

https://github.com/zephyrproject-rtos/zephyr/blob/127cb9edb6c6594e2f09dea7efb335db3cb1f43a/include/zephyr/arch/arch_inlines.h#L35-L37

https://github.com/zephyrproject-rtos/zephyr/blob/127cb9edb6c6594e2f09dea7efb335db3cb1f43a/include/zephyr/arch/cpu.h#L34-L36

Another PR is required if normalization is required

aescolar replied:
We support out of tree architectures, SOCs, and boards. For out of tree architectures, because they are so rare, out of tree support gets a bit broken every now and then with this kind of changes. This can be worked around/hacked out of tree. But that does not mean we should just keep making it worse for no reason.
I don't think there is any benefit to add this check and error here. So I'd still ask you to not add it.

ycsin replied:
Fair enough, added 2 more commits that will make these headers more consistent

ycsin committed:
arch: rename all esf struct to `struct arch_esf`

Rename every architecture's esf struct to `struct esf`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

Thalley commented:
CI issues seems to stem from this https://github.com/zephyrproject-rtos/zephyr/issues/73603

Thalley reviewed (commented):
Do the changes in this PR (specifically the ones that modify the API in `include/zephyr`), constitute a breaking stable API change? 

Does the PR need to go through https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes ?

ycsin commented:
> Do the changes in this PR (specifically the ones that modify the API in `include/zephyr`), constitute a breaking stable API change?
> 
> Does the PR need to go through https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes ?

Maybe - I'm not sure how to deprecate a named struct though, probably wrap its declaration with a Kconfig and generate warning message when the Kconfig is enabled?

ycsin committed:
arch: remove the use of `z_arch_esf_t` completely from internal

Created `GEN_OFFSET_STRUCT` & `GEN_NAMED_OFFSET_STRUCT` that
works for `struct`, and remove the use of `z_arch_esf_t`
completely.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
doc: releases: update migration guide for the `struct arch_esf`

Add a note about the introduction of `struct arch_esf` and the
deprecation of `z_arch_esf_t`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Removed all usage of `z_arch_esf_t` from zephyr base.

nashif commented:
> Do the changes in this PR (specifically the ones that modify the API in `include/zephyr`), constitute a breaking stable API change?
> 
> Does the PR need to go through [docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes](https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes) ?

well well, z_ has been always used for private APIs, variables, obviously this was wrong used here (and we have a few others of this class), however, if this is deprecagted correctly, it is not a breaking API change and does not follow the process IMO, just the migration documentation.

This change is very welcome, wanted to do that for a long time, never got to it...

ycsin committed:
manifest: update `sof` git revision

Update the git revision of the `sof` module.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
Let's see if `sof` is the only external module affected by this change

andyross reviewed (dismissed):
This looks perfect, thanks.

ycsin commented:
Updated to git revision of `sof` to point at the hash after the merge of https://github.com/zephyrproject-rtos/sof/pull/46

aescolar commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -59,6 +59,11 @@ Build System
 Kernel
 ******
 
+* All architectures are now required to define the new ``struct arch_esf``, which describes the members
```
Please add a 
```
(:github:`73593`).
```

ycsin replied:
fixes

ycsin committed:
include: optimize the order of includes for x86

x86 32bit defines `CONFIG_X86` while its 64bit counterpart
defines an additional `CONFIG_X86_64`, by reordering the
include order we can make it look a bit cleaner.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
include: stop breaking the compilation of OOT architectures

Zephyr support out-of-tree architectures so we shouldn't
throw error for archs not listed here.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
ping @andyross could you please take another look? Thanks

Thalley commented:
Thanks for the quick fix @ycsin 

One thing I can see today is that now `zephyr/include/zephyr/arch/arm/cortex_m/exception.h` may be included before `sys_define_gpr_with_alias` has been defined. 

This effectively means that the new generic `exception.h` header cannot be included for ARM before `zephyr/include/zephyr/arch/arm/arch.h` has been included, as `zephyr/include/zephyr/arch/arm/arch.h` defines the `sys_define_gpr_with_alias` macro. 

I almost expect that if we fix the above, yet another include issue will appear. These type of include issues will continue to appear unless verified by a tool (attempting to fix or verify manually will be nearly impossible).

Thalley commented:
> Thanks for the quick fix @ycsin
> 
> One thing I can see today is that now `zephyr/include/zephyr/arch/arm/cortex_m/exception.h` may be included before `sys_define_gpr_with_alias` has been defined.
> 
> This effectively means that the new generic `exception.h` header cannot be included for ARM before `zephyr/include/zephyr/arch/arm/arch.h` has been included, as `zephyr/include/zephyr/arch/arm/arch.h` defines the `sys_define_gpr_with_alias` macro.
> 
> I almost expect that if we fix the above, yet another include issue will appear. These type of include issues will continue to appear unless verified by a tool (attempting to fix or verify manually will be nearly impossible).

@ycsin are you looking into this? Otherwise I'll go ahead and create another issue :) 

If you are not already, you may want to use a tool like the `clang` `includer-cleaner` or `clangd` to catch future issues.

ycsin commented:
> @ycsin are you looking into this? Otherwise I'll go ahead and create another issue :)

Not looking into this, please feel free to create another issue

Thalley commented:
> > @ycsin are you looking into this? Otherwise I'll go ahead and create another issue :)
> 
> Not looking into this, please feel free to create another issue

The fix for this was quite simple, so decided to take a look at fixing it myself: https://github.com/zephyrproject-rtos/zephyr/pull/74106
